content/posts/csv-boilerplate-for-go.md

```diff
@@ -0,0 +1,307 @@
+3. The result is not very robust when it comes to bad CSV input and doesn't handle problems such as missing, unexpected, or misordered columns or headers.
+4. Error messages often end up missing important information such as row number, column name, etc..
+5. How do I write good tests for this that will be easy to maintain in the future?
+
+One way to deal with this is to use a high level library, and you should certainly consider that. But the reality is often an ugly mess. You might find that the library of your choice doesn't handle a particular edge case, e.g. more than one header row being present (my bank loves putting in 6 header lines). Or the error message you're getting are not helpful enough for your users. Or the error handling is too strict. Or the reflection based mapping is too magic. Or the performance sucks. I could go on and on, but I think you get the point - dealing with CSV is nasty and calling it a format is a bit of an insult to the idea of a format. From my point of view, writing a library for this problem is simply not possible without creating a horribly overengineered mess or ignoring many edge cases.
+
+So in the end it seems like you have to chose between two suboptimal options. You can try out a bunch of libraries and use and/or fork the one closest to your needs. Or you can roll your own boilerplate for the millionth time, only to look back at it with horrible disgust shortly after.
+
+Your choice should probably depend on the crazyness of your input, and I won't be able to make it for you. But I've decided that I want to make the boilerplate route a bit less annoying for myself in the future. To accomplish this I've created a [boilerplate template](https://github.com/felixge/dump/blob/master/csv-boilerplate/main.go) that I'll copy and paste for my future needs, adjusting it as needed. For the rest of the blog post I want to walk you through my thought process in designing it.
+
+To begin, I decided I wanted to use a real world CSV file as my example that has a bit of the typical uglyness one has to deal with. I quickly [found](https://perso.telecom-paristech.fr/eagan/class/igr204/datasets) a file called [cars.csv](https://github.com/felixge/dump/blob/master/csv-boilerplate/cars.csv) which looks like this:
+
+```csv
+Car;MPG;Cylinders;Displacement;Horsepower;Weight;Acceleration;Model;Origin
+STRING;DOUBLE;INT;DOUBLE;DOUBLE;DOUBLE;DOUBLE;INT;CAT
+Chevrolet Chevelle Malibu;18.0;8;307.0;130.0;3504.;12.0;70;US
+Buick Skylark 320;15.0;8;350.0;165.0;3693.;11.5;70;US
+Plymouth Satellite;18.0;8;318.0;150.0;3436.;11.0;70;US
+...
+```
+
+As you can see, it uses semicolons instead of commas as a separator, and it has a second header line that seems to contain type information for whatever reason. Not great, not terrible, but pretty typical for what's out there.
```

Comment:
😛 

```suggestion
As you can see, there are already two ugly bits to be examined here: semicolons instead of commas as a separator, and a second header line that seems to contain type information for whatever reason. Not great, not terrible, but pretty typical for what's out there.
```